pkg/provider/vault/vault.go

```diff
@@ -0,0 +1,391 @@
+	_ provider.SecretsClient = &client{}
+)
+
+const (
+	serviceAccTokenPath = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+
+	errVaultStore     = "received invalid Vault SecretStore resource: %w"
+	errVaultClient    = "cannot setup new vault client: %w"
+	errVaultCert      = "cannot set Vault CA certificate: %w"
+	errReadSecret     = "cannot read secret data from Vault: %w"
+	errAuthFormat     = "cannot initialize Vault client: no valid auth method specified: %w"
+	errVaultData      = "cannot parse Vault response data: %w"
+	errVaultToken     = "cannot parse Vault authentication token: %w"
+	errVaultReqParams = "cannot set Vault request parameters: %w"
+	errVaultRequest   = "error from Vault request: %w"
+	errVaultResponse  = "cannot parse Vault response: %w"
+	errServiceAccount = "cannot read Kubernetes service account token from file system: %w"
+
+	errGetKubeSecret = "cannot get Kubernetes secret %q: %w"
+	errSecretKeyFmt  = "cannot find secret data for key: %q"
+)
+
+type Client interface {
+	NewRequest(method, requestPath string) *vault.Request
+	RawRequestWithContext(ctx context.Context, r *vault.Request) (*vault.Response, error)
+	SetToken(v string)
+	SetNamespace(namespace string)
+}
+
+type client struct {
+	kube      kclient.Client
+	store     *esv1alpha1.VaultProvider
+	log       logr.Logger
+	client    Client
+	namespace string
+	storeKind string
+}
+
+func init() {
+	schema.Register(&connector{
+		newVaultClient: newVaultClient,
+	}, &esv1alpha1.SecretStoreProvider{
+		Vault: &esv1alpha1.VaultProvider{},
+	})
+}
+
+func newVaultClient(c *vault.Config) (Client, error) {
+	return vault.NewClient(c)
+}
+
+type connector struct {
+	newVaultClient func(c *vault.Config) (Client, error)
+}
+
+func (c *connector) NewClient(ctx context.Context, store esv1alpha1.GenericStore, kube kclient.Client, namespace string) (provider.SecretsClient, error) {
+	storeSpec := store.GetSpec()
+	if storeSpec == nil || storeSpec.Provider == nil || storeSpec.Provider.Vault == nil {
+		return nil, errors.New(errVaultStore)
+	}
+	vaultSpec := storeSpec.Provider.Vault
+
+	vStore := &client{
+		kube:      kube,
+		store:     vaultSpec,
+		log:       ctrl.Log.WithName("provider").WithName("vault"),
+		namespace: namespace,
+		storeKind: store.GetObjectKind().GroupVersionKind().Kind,
+	}
+
+	cfg, err := vStore.newConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := c.newVaultClient(cfg)
+	if err != nil {
+		return nil, fmt.Errorf(errVaultClient, err)
+	}
+
+	if vaultSpec.Namespace != nil {
+		client.SetNamespace(*vaultSpec.Namespace)
+	}
+
+	if err := vStore.setAuth(ctx, client); err != nil {
+		return nil, err
+	}
+
+	vStore.client = client
+	return vStore, nil
+}
+
+func (v *client) GetSecret(ctx context.Context, ref esv1alpha1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	data, err := v.readSecret(ctx, ref.Key, ref.Version)
+	if err != nil {
+		return nil, err
+	}
+	value, exists := data[ref.Property]
```

Comment:
Should we use `ref.Key` here instead of `ref.Property`?

From my perspective, the SecretStore and ExternalSecret should look like:

SecretStore:
```yaml
...
spec:
  controller: default
  provider:
    vault:
      version: v1
      auth:
        tokenSecretRef:
          name: vault
          key: root-token

      server: http://vault.default.svc.cluster.local:8200

      path: kv/my-secret
```

ExternalSecret:
```yaml
...
spec:
  secretStoreRef: 
    name: vault-sample

  target:
    name: vault-secret-to-be-created

  data:
  - secretKey: vault-key-to-be-created
    remoteRef:
      # this is the key I'm looking for from the `data` field
      # this key has no "property", that's why I think we could use the key field
      key: name
```